In [117]:
def train_imat(method):
    method = str(method)
    !time java -jar ../../lib/RankLib-2.12.jar -train ../../data/task3_train.txt -tts 0.2 -ranker {method} -metric2t NDCG@20 -sparse -silent

In [118]:
methods = [(0, "MART"), (1, "RankNet"), (2, "RankBoost"), (3, "AdaRank"), (4, "Coordinate Ascent"), (6, "LambdaMART"), (7, "ListNet"), (8, "Random Forests")]

In [119]:
for method, name in methods:
    print("Method: " + name)
    train_imat(method)

Method: MART

Discard orig. features
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	MART
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] MART's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7428
208.13user 1.77system 0:52.50elapsed 399%CPU (0avgtext+0avgdata 562868maxresident)k
0inputs+216outputs (0major+155879minor)pagefaults 0swaps
Method: RankNet

Discard orig. features
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	RankNet
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] RankNet's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]      

Method: MART  
NDCG@20 on test data: 0.7428 
208.13  

Method: RankNet  
NDCG@20 on test data: 0.6684
152.85

Method: RankBoost  
NDCG@20 on test data: 0.7189 
285.04

Method: AdaRank  
NDCG@20 on test data: 0.7137  
72.18

Method: Coordinate Ascent  
NDCG@20 on test data: 0.7292  
378.10  

Method: LambdaMART  
NDCG@20 on test data: 0.7422 
201.39

Method: ListNet  
NDCG@20 on test data: 0.7272  
371.91 

Method: Random Forests  
NDCG@20 on test data: 0.7451
603.21

Интересны MART и Random Forests, если пытаться получить лучший результат

# MART

In [120]:
!time java -jar ../../lib/RankLib-2.12.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 0 -tree 3000 -leaf 100 -metric2t NDCG@20 -sparse -silent


Discard orig. features
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	MART
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] MART's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7367
2997.11user 41.98system 12:01.53elapsed 421%CPU (0avgtext+0avgdata 586932maxresident)k
0inputs+1920outputs (0major+219208minor)pagefaults 0swaps


In [62]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 0 -tree 3000 -leaf 10 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	MART
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] MART's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7355
1004.63user 1.96system 12:55.99elapsed 129%CPU (0avgtext+0avgdata 594688maxresident)k
0inputs+1576outputs (0major+159373minor)pagefaults 0swaps


In [121]:
!time java -jar ../../lib/RankLib-2.12.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 0 -tree 500 -leaf 10 -metric2t NDCG@20 -sparse -silent


Discard orig. features
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	MART
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] MART's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7432
101.04user 0.94system 0:25.62elapsed 397%CPU (0avgtext+0avgdata 582440maxresident)k
0inputs+128outputs (0major+159948minor)pagefaults 0swaps


In [122]:
!time java -jar ../../lib/RankLib-2.12.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 0 -tree 500 -leaf 20 -metric2t NDCG@20 -sparse -silent


Discard orig. features
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	MART
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] MART's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7426
179.48user 1.67system 0:42.03elapsed 430%CPU (0avgtext+0avgdata 578044maxresident)k
0inputs+232outputs (0major+161331minor)pagefaults 0swaps


In [123]:
!time java -jar ../../lib/RankLib-2.12.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 0 -tree 300 -leaf 300 -metric2t NDCG@20 -sparse -silent


Discard orig. features
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	MART
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] MART's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7372
498.06user 14.44system 1:52.26elapsed 456%CPU (0avgtext+0avgdata 1793472maxresident)k
0inputs+568outputs (0major+1015846minor)pagefaults 0swaps


In [133]:
!time java -jar ../../lib/RankLib-2.12.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 0 -tree 1000 -leaf 200 -metric2t NDCG@20 -sparse -silent


Discard orig. features
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	MART
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] MART's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7382
1519.05user 30.49system 5:48.94elapsed 444%CPU (0avgtext+0avgdata 1205060maxresident)k
0inputs+656outputs (0major+623592minor)pagefaults 0swaps


# Random Forests

In [63]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 300 -leaf 100 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.7346    | 
b[2]      | 0.7221    | 
b[3]      | 0.7276    | 
b[4]      | 0.729     | 
b[5]      | 0.7353    | 
b[6]      | 0.7276    | 
b[7]      | 0.7396    | 
b[8]      | 0.733     | 
b[9]      | 0.7113    | 
b[10]     | 0.7182    | 
b[11]     | 0.7348    | 
b[12]     | 0.736     | 
b[13]     | 0.736     | 
b[14]     | 0.7321    | 
b[15]     | 0.7505    | 
b[16]     | 0.7574    | 
b[17]     | 0.7338    | 
b[18]     | 0.7314    | 
b[19]     | 0.7416    | 
b[20]     | 0.7375    | 
b[21]     | 0.7446    | 
b[22]     | 0.7328    | 
b[23]   

In [64]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 300 -leaf 10 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.6885    | 
b[2]      | 0.7005    | 
b[3]      | 0.7048    | 
b[4]      | 0.6928    | 
b[5]      | 0.6673    | 
b[6]      | 0.6898    | 
b[7]      | 0.6897    | 
b[8]      | 0.693     | 
b[9]      | 0.6906    | 
b[10]     | 0.685     | 
b[11]     | 0.6892    | 
b[12]     | 0.6813    | 
b[13]     | 0.7011    | 
b[14]     | 0.6845    | 
b[15]     | 0.6905    | 
b[16]     | 0.6883    | 
b[17]     | 0.689     | 
b[18]     | 0.6821    | 
b[19]     | 0.6966    | 
b[20]     | 0.703     | 
b[21]     | 0.6856    | 
b[22]     | 0.6831    | 
b[23]   

In [66]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 1000 -leaf 20 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.7055    | 
b[2]      | 0.7087    | 
b[3]      | 0.6953    | 
b[4]      | 0.6842    | 
b[5]      | 0.7084    | 
b[6]      | 0.7057    | 
b[7]      | 0.672     | 
b[8]      | 0.7054    | 
b[9]      | 0.7035    | 
b[10]     | 0.6827    | 
b[11]     | 0.6982    | 
b[12]     | 0.7006    | 
b[13]     | 0.7053    | 
b[14]     | 0.6921    | 
b[15]     | 0.6988    | 
b[16]     | 0.7072    | 
b[17]     | 0.7107    | 
b[18]     | 0.6896    | 
b[19]     | 0.7006    | 
b[20]     | 0.6944    | 
b[21]     | 0.6894    | 
b[22]     | 0.7104    | 
b[23]   

b[312]    | 0.7102    | 
b[313]    | 0.7024    | 
b[314]    | 0.6975    | 
b[315]    | 0.7071    | 
b[316]    | 0.7019    | 
b[317]    | 0.6889    | 
b[318]    | 0.7014    | 
b[319]    | 0.6989    | 
b[320]    | 0.6925    | 
b[321]    | 0.6912    | 
b[322]    | 0.6944    | 
b[323]    | 0.6929    | 
b[324]    | 0.7028    | 
b[325]    | 0.6863    | 
b[326]    | 0.7107    | 
b[327]    | 0.7129    | 
b[328]    | 0.696     | 
b[329]    | 0.7006    | 
b[330]    | 0.6984    | 
b[331]    | 0.6952    | 
b[332]    | 0.6945    | 
b[333]    | 0.6927    | 
b[334]    | 0.724     | 
b[335]    | 0.7133    | 
b[336]    | 0.6905    | 
b[337]    | 0.6899    | 
b[338]    | 0.6888    | 
b[339]    | 0.714     | 
b[340]    | 0.7106    | 
b[341]    | 0.7037    | 
b[342]    | 0.7067    | 
b[343]    | 0.7029    | 
b[344]    | 0.7076    | 
b[345]    | 0.7088    | 
b[346]    | 0.6939    | 
b[347]    | 0.6981    | 
b[348]    | 0.6961    | 
b[349]    | 0.7025    | 
b[350]    | 0.7012    | 
b[351]    | 0.7153    | 


b[640]    | 0.7085    | 
b[641]    | 0.6849    | 
b[642]    | 0.7051    | 
b[643]    | 0.7113    | 
b[644]    | 0.7012    | 
b[645]    | 0.7085    | 
b[646]    | 0.6897    | 
b[647]    | 0.7076    | 
b[648]    | 0.7052    | 
b[649]    | 0.7003    | 
b[650]    | 0.7042    | 
b[651]    | 0.6968    | 
b[652]    | 0.7037    | 
b[653]    | 0.6993    | 
b[654]    | 0.6953    | 
b[655]    | 0.6861    | 
b[656]    | 0.697     | 
b[657]    | 0.7035    | 
b[658]    | 0.7218    | 
b[659]    | 0.7068    | 
b[660]    | 0.7165    | 
b[661]    | 0.6976    | 
b[662]    | 0.7031    | 
b[663]    | 0.7069    | 
b[664]    | 0.6811    | 
b[665]    | 0.7024    | 
b[666]    | 0.6989    | 
b[667]    | 0.6884    | 
b[668]    | 0.7039    | 
b[669]    | 0.6918    | 
b[670]    | 0.7191    | 
b[671]    | 0.6863    | 
b[672]    | 0.7005    | 
b[673]    | 0.7076    | 
b[674]    | 0.7035    | 
b[675]    | 0.7132    | 
b[676]    | 0.7063    | 
b[677]    | 0.7023    | 
b[678]    | 0.702     | 
b[679]    | 0.683     | 


b[968]    | 0.6951    | 
b[969]    | 0.6969    | 
b[970]    | 0.7041    | 
b[971]    | 0.7141    | 
b[972]    | 0.6967    | 
b[973]    | 0.7127    | 
b[974]    | 0.6987    | 
b[975]    | 0.7032    | 
b[976]    | 0.697     | 
b[977]    | 0.703     | 
b[978]    | 0.701     | 
b[979]    | 0.7031    | 
b[980]    | 0.7064    | 
b[981]    | 0.6885    | 
b[982]    | 0.7015    | 
b[983]    | 0.71      | 
b[984]    | 0.7199    | 
b[985]    | 0.7084    | 
b[986]    | 0.7053    | 
b[987]    | 0.6924    | 
b[988]    | 0.6962    | 
b[989]    | 0.6951    | 
b[990]    | 0.7097    | 
b[991]    | 0.6885    | 
b[992]    | 0.7007    | 
b[993]    | 0.7122    | 
b[994]    | 0.6952    | 
b[995]    | 0.7043    | 
b[996]    | 0.7023    | 
b[997]    | 0.7112    | 
b[998]    | 0.7021    | 
b[999]    | 0.6997    | 
b[1000]   | 0.7065    | 
------------------------------------
Finished sucessfully.
NDCG@20 on training data: 0.7352
------------------------------------
NDCG@20 on test data: 0.7376
2699.91user 36.40

In [68]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 300 -leaf 300 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.7815    | 
b[2]      | 0.7733    | 
b[3]      | 0.779     | 
b[4]      | 0.7726    | 
b[5]      | 0.7816    | 
b[6]      | 0.7693    | 
b[7]      | 0.7813    | 
b[8]      | 0.7677    | 
b[9]      | 0.7775    | 
b[10]     | 0.7818    | 
b[11]     | 0.7927    | 
b[12]     | 0.7945    | 
b[13]     | 0.7613    | 
b[14]     | 0.7819    | 
b[15]     | 0.7777    | 
b[16]     | 0.7932    | 
b[17]     | 0.7751    | 
b[18]     | 0.7723    | 
b[19]     | 0.7802    | 
b[20]     | 0.7706    | 
b[21]     | 0.7689    | 
b[22]     | 0.7648    | 
b[23]   

In [70]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 1000 -leaf 300 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.7901    | 
b[2]      | 0.7705    | 
b[3]      | 0.7796    | 
b[4]      | 0.771     | 
b[5]      | 0.7703    | 
b[6]      | 0.789     | 
b[7]      | 0.7894    | 
b[8]      | 0.7879    | 
b[9]      | 0.7823    | 
b[10]     | 0.8002    | 
b[11]     | 0.7807    | 
b[12]     | 0.7887    | 
b[13]     | 0.7713    | 
b[14]     | 0.7638    | 
b[15]     | 0.7999    | 
b[16]     | 0.7745    | 
b[17]     | 0.7774    | 
b[18]     | 0.7669    | 
b[19]     | 0.7681    | 
b[20]     | 0.7763    | 
b[21]     | 0.7743    | 
b[22]     | 0.7826    | 
b[23]   

b[312]    | 0.7861    | 
b[313]    | 0.783     | 
b[314]    | 0.7669    | 
b[315]    | 0.7925    | 
b[316]    | 0.7841    | 
b[317]    | 0.7743    | 
b[318]    | 0.7736    | 
b[319]    | 0.7867    | 
b[320]    | 0.7955    | 
b[321]    | 0.7752    | 
b[322]    | 0.7664    | 
b[323]    | 0.7688    | 
b[324]    | 0.7731    | 
b[325]    | 0.7747    | 
b[326]    | 0.7774    | 
b[327]    | 0.7742    | 
b[328]    | 0.7833    | 
b[329]    | 0.7795    | 
b[330]    | 0.7712    | 
b[331]    | 0.7908    | 
b[332]    | 0.768     | 
b[333]    | 0.7671    | 
b[334]    | 0.7781    | 
b[335]    | 0.7804    | 
b[336]    | 0.7913    | 
b[337]    | 0.7851    | 
b[338]    | 0.7861    | 
b[339]    | 0.7937    | 
b[340]    | 0.793     | 
b[341]    | 0.7855    | 
b[342]    | 0.7932    | 
b[343]    | 0.7861    | 
b[344]    | 0.79      | 
b[345]    | 0.7639    | 
b[346]    | 0.7655    | 
b[347]    | 0.7773    | 
b[348]    | 0.7723    | 
b[349]    | 0.7716    | 
b[350]    | 0.797     | 
b[351]    | 0.7768    | 


b[640]    | 0.7907    | 
b[641]    | 0.7611    | 
b[642]    | 0.7761    | 
b[643]    | 0.7957    | 
b[644]    | 0.7879    | 
b[645]    | 0.777     | 
b[646]    | 0.7698    | 
b[647]    | 0.7809    | 
b[648]    | 0.7578    | 
b[649]    | 0.7869    | 
b[650]    | 0.7678    | 
b[651]    | 0.7833    | 
b[652]    | 0.784     | 
b[653]    | 0.7785    | 
b[654]    | 0.7743    | 
b[655]    | 0.7821    | 
b[656]    | 0.7663    | 
b[657]    | 0.7792    | 
b[658]    | 0.7828    | 
b[659]    | 0.776     | 
b[660]    | 0.7728    | 
b[661]    | 0.793     | 
b[662]    | 0.7709    | 
b[663]    | 0.7863    | 
b[664]    | 0.7859    | 
b[665]    | 0.797     | 
b[666]    | 0.7934    | 
b[667]    | 0.7605    | 
b[668]    | 0.7782    | 
b[669]    | 0.7933    | 
b[670]    | 0.7685    | 
b[671]    | 0.7663    | 
b[672]    | 0.7624    | 
b[673]    | 0.7811    | 
b[674]    | 0.7814    | 
b[675]    | 0.7762    | 
b[676]    | 0.7827    | 
b[677]    | 0.7964    | 
b[678]    | 0.7835    | 
b[679]    | 0.7757    | 


b[968]    | 0.761     | 
b[969]    | 0.7648    | 
b[970]    | 0.7726    | 
b[971]    | 0.7686    | 
b[972]    | 0.7839    | 
b[973]    | 0.7672    | 
b[974]    | 0.7609    | 
b[975]    | 0.7704    | 
b[976]    | 0.7747    | 
b[977]    | 0.7837    | 
b[978]    | 0.7727    | 
b[979]    | 0.7956    | 
b[980]    | 0.778     | 
b[981]    | 0.7893    | 
b[982]    | 0.7781    | 
b[983]    | 0.7763    | 
b[984]    | 0.7702    | 
b[985]    | 0.7831    | 
b[986]    | 0.7756    | 
b[987]    | 0.7832    | 
b[988]    | 0.793     | 
b[989]    | 0.783     | 
b[990]    | 0.7825    | 
b[991]    | 0.7648    | 
b[992]    | 0.7785    | 
b[993]    | 0.777     | 
b[994]    | 0.766     | 
b[995]    | 0.7792    | 
b[996]    | 0.7903    | 
b[997]    | 0.7822    | 
b[998]    | 0.7866    | 
b[999]    | 0.7855    | 
b[1000]   | 0.7761    | 
------------------------------------
Finished sucessfully.
NDCG@20 on training data: 0.8184
------------------------------------
NDCG@20 on test data: 0.7482
4128.88user 182.5

## Пни

In [78]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 1000 -leaf 10 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.6761    | 
b[2]      | 0.7055    | 
b[3]      | 0.6751    | 
b[4]      | 0.6981    | 
b[5]      | 0.6939    | 
b[6]      | 0.6885    | 
b[7]      | 0.704     | 
b[8]      | 0.7       | 
b[9]      | 0.6964    | 
b[10]     | 0.6818    | 
b[11]     | 0.6918    | 
b[12]     | 0.6828    | 
b[13]     | 0.6885    | 
b[14]     | 0.6847    | 
b[15]     | 0.6765    | 
b[16]     | 0.6914    | 
b[17]     | 0.6917    | 
b[18]     | 0.6819    | 
b[19]     | 0.7       | 
b[20]     | 0.6827    | 
b[21]     | 0.6863    | 
b[22]     | 0.695     | 
b[23]   

b[312]    | 0.7033    | 
b[313]    | 0.693     | 
b[314]    | 0.6805    | 
b[315]    | 0.676     | 
b[316]    | 0.6897    | 
b[317]    | 0.6716    | 
b[318]    | 0.6624    | 
b[319]    | 0.6937    | 
b[320]    | 0.683     | 
b[321]    | 0.6866    | 
b[322]    | 0.6791    | 
b[323]    | 0.6834    | 
b[324]    | 0.696     | 
b[325]    | 0.6744    | 
b[326]    | 0.6905    | 
b[327]    | 0.6788    | 
b[328]    | 0.67      | 
b[329]    | 0.6958    | 
b[330]    | 0.6716    | 
b[331]    | 0.6826    | 
b[332]    | 0.6921    | 
b[333]    | 0.6862    | 
b[334]    | 0.6862    | 
b[335]    | 0.6901    | 
b[336]    | 0.6903    | 
b[337]    | 0.6886    | 
b[338]    | 0.6996    | 
b[339]    | 0.6666    | 
b[340]    | 0.662     | 
b[341]    | 0.6906    | 
b[342]    | 0.701     | 
b[343]    | 0.6759    | 
b[344]    | 0.6689    | 
b[345]    | 0.6784    | 
b[346]    | 0.6781    | 
b[347]    | 0.6778    | 
b[348]    | 0.6854    | 
b[349]    | 0.6679    | 
b[350]    | 0.6969    | 
b[351]    | 0.6933    | 


b[640]    | 0.6856    | 
b[641]    | 0.7008    | 
b[642]    | 0.6956    | 
b[643]    | 0.6855    | 
b[644]    | 0.6763    | 
b[645]    | 0.6972    | 
b[646]    | 0.6976    | 
b[647]    | 0.6828    | 
b[648]    | 0.6916    | 
b[649]    | 0.6861    | 
b[650]    | 0.6904    | 
b[651]    | 0.6978    | 
b[652]    | 0.7056    | 
b[653]    | 0.6772    | 
b[654]    | 0.6777    | 
b[655]    | 0.6851    | 
b[656]    | 0.693     | 
b[657]    | 0.6827    | 
b[658]    | 0.6753    | 
b[659]    | 0.7164    | 
b[660]    | 0.6872    | 
b[661]    | 0.6747    | 
b[662]    | 0.6932    | 
b[663]    | 0.7073    | 
b[664]    | 0.6842    | 
b[665]    | 0.6966    | 
b[666]    | 0.6952    | 
b[667]    | 0.7045    | 
b[668]    | 0.7043    | 
b[669]    | 0.7011    | 
b[670]    | 0.6844    | 
b[671]    | 0.6937    | 
b[672]    | 0.6854    | 
b[673]    | 0.6854    | 
b[674]    | 0.6997    | 
b[675]    | 0.6832    | 
b[676]    | 0.6879    | 
b[677]    | 0.7167    | 
b[678]    | 0.6889    | 
b[679]    | 0.6983    | 


b[968]    | 0.6872    | 
b[969]    | 0.6846    | 
b[970]    | 0.6895    | 
b[971]    | 0.677     | 
b[972]    | 0.6773    | 
b[973]    | 0.6951    | 
b[974]    | 0.6755    | 
b[975]    | 0.6837    | 
b[976]    | 0.6875    | 
b[977]    | 0.6828    | 
b[978]    | 0.6911    | 
b[979]    | 0.6862    | 
b[980]    | 0.6822    | 
b[981]    | 0.6792    | 
b[982]    | 0.6636    | 
b[983]    | 0.7023    | 
b[984]    | 0.6996    | 
b[985]    | 0.6685    | 
b[986]    | 0.6782    | 
b[987]    | 0.6742    | 
b[988]    | 0.6878    | 
b[989]    | 0.6914    | 
b[990]    | 0.6868    | 
b[991]    | 0.6897    | 
b[992]    | 0.6808    | 
b[993]    | 0.6879    | 
b[994]    | 0.6899    | 
b[995]    | 0.7036    | 
b[996]    | 0.7117    | 
b[997]    | 0.7086    | 
b[998]    | 0.6803    | 
b[999]    | 0.684     | 
b[1000]   | 0.6844    | 
------------------------------------
Finished sucessfully.
NDCG@20 on training data: 0.7254
------------------------------------
NDCG@20 on test data: 0.7337
2529.40user 35.56

## Пеньки

In [79]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 1000 -leaf 5 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.6806    | 
b[2]      | 0.6889    | 
b[3]      | 0.6736    | 
b[4]      | 0.6719    | 
b[5]      | 0.6883    | 
b[6]      | 0.6814    | 
b[7]      | 0.6596    | 
b[8]      | 0.6629    | 
b[9]      | 0.67      | 
b[10]     | 0.6784    | 
b[11]     | 0.6733    | 
b[12]     | 0.6653    | 
b[13]     | 0.6761    | 
b[14]     | 0.667     | 
b[15]     | 0.6849    | 
b[16]     | 0.6786    | 
b[17]     | 0.6687    | 
b[18]     | 0.6654    | 
b[19]     | 0.6496    | 
b[20]     | 0.6924    | 
b[21]     | 0.6854    | 
b[22]     | 0.671     | 
b[23]   

b[312]    | 0.6634    | 
b[313]    | 0.6875    | 
b[314]    | 0.6671    | 
b[315]    | 0.6711    | 
b[316]    | 0.6738    | 
b[317]    | 0.6754    | 
b[318]    | 0.6754    | 
b[319]    | 0.6734    | 
b[320]    | 0.6889    | 
b[321]    | 0.6705    | 
b[322]    | 0.6854    | 
b[323]    | 0.6804    | 
b[324]    | 0.6686    | 
b[325]    | 0.6818    | 
b[326]    | 0.672     | 
b[327]    | 0.68      | 
b[328]    | 0.6653    | 
b[329]    | 0.6625    | 
b[330]    | 0.6557    | 
b[331]    | 0.6633    | 
b[332]    | 0.6606    | 
b[333]    | 0.666     | 
b[334]    | 0.686     | 
b[335]    | 0.6764    | 
b[336]    | 0.6767    | 
b[337]    | 0.6796    | 
b[338]    | 0.6879    | 
b[339]    | 0.6974    | 
b[340]    | 0.6735    | 
b[341]    | 0.6692    | 
b[342]    | 0.6799    | 
b[343]    | 0.6601    | 
b[344]    | 0.689     | 
b[345]    | 0.6765    | 
b[346]    | 0.6749    | 
b[347]    | 0.6506    | 
b[348]    | 0.6869    | 
b[349]    | 0.6752    | 
b[350]    | 0.665     | 
b[351]    | 0.6745    | 


b[640]    | 0.6836    | 
b[641]    | 0.6642    | 
b[642]    | 0.674     | 
b[643]    | 0.659     | 
b[644]    | 0.689     | 
b[645]    | 0.6956    | 
b[646]    | 0.6789    | 
b[647]    | 0.6706    | 
b[648]    | 0.6818    | 
b[649]    | 0.6698    | 
b[650]    | 0.6635    | 
b[651]    | 0.6773    | 
b[652]    | 0.6696    | 
b[653]    | 0.695     | 
b[654]    | 0.6679    | 
b[655]    | 0.6726    | 
b[656]    | 0.6966    | 
b[657]    | 0.6677    | 
b[658]    | 0.6761    | 
b[659]    | 0.6739    | 
b[660]    | 0.6926    | 
b[661]    | 0.6878    | 
b[662]    | 0.6708    | 
b[663]    | 0.6785    | 
b[664]    | 0.6662    | 
b[665]    | 0.6671    | 
b[666]    | 0.6666    | 
b[667]    | 0.6676    | 
b[668]    | 0.6867    | 
b[669]    | 0.6736    | 
b[670]    | 0.6608    | 
b[671]    | 0.6503    | 
b[672]    | 0.6877    | 
b[673]    | 0.6569    | 
b[674]    | 0.6751    | 
b[675]    | 0.6726    | 
b[676]    | 0.6793    | 
b[677]    | 0.6719    | 
b[678]    | 0.6817    | 
b[679]    | 0.6854    | 


b[968]    | 0.676     | 
b[969]    | 0.6567    | 
b[970]    | 0.6773    | 
b[971]    | 0.6749    | 
b[972]    | 0.6724    | 
b[973]    | 0.6587    | 
b[974]    | 0.6741    | 
b[975]    | 0.6897    | 
b[976]    | 0.6872    | 
b[977]    | 0.66      | 
b[978]    | 0.6701    | 
b[979]    | 0.6773    | 
b[980]    | 0.6909    | 
b[981]    | 0.661     | 
b[982]    | 0.6653    | 
b[983]    | 0.6847    | 
b[984]    | 0.6801    | 
b[985]    | 0.6598    | 
b[986]    | 0.6617    | 
b[987]    | 0.6805    | 
b[988]    | 0.6782    | 
b[989]    | 0.6814    | 
b[990]    | 0.6907    | 
b[991]    | 0.6885    | 
b[992]    | 0.6759    | 
b[993]    | 0.6527    | 
b[994]    | 0.6679    | 
b[995]    | 0.6735    | 
b[996]    | 0.6789    | 
b[997]    | 0.6834    | 
b[998]    | 0.6708    | 
b[999]    | 0.6686    | 
b[1000]   | 0.6936    | 
------------------------------------
Finished sucessfully.
NDCG@20 on training data: 0.7159
------------------------------------
NDCG@20 on test data: 0.728
2154.50user 32.99s

In [81]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 100 -leaf 50 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.7109    | 
b[2]      | 0.7221    | 
b[3]      | 0.7198    | 
b[4]      | 0.7143    | 
b[5]      | 0.7319    | 
b[6]      | 0.7426    | 
b[7]      | 0.7258    | 
b[8]      | 0.7088    | 
b[9]      | 0.7209    | 
b[10]     | 0.7123    | 
b[11]     | 0.7328    | 
b[12]     | 0.7253    | 
b[13]     | 0.7085    | 
b[14]     | 0.7201    | 
b[15]     | 0.7188    | 
b[16]     | 0.7256    | 
b[17]     | 0.7249    | 
b[18]     | 0.7156    | 
b[19]     | 0.7279    | 
b[20]     | 0.7122    | 
b[21]     | 0.7228    | 
b[22]     | 0.7179    | 
b[23]   

In [82]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 100 -leaf 300 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.7999    | 
b[2]      | 0.7729    | 
b[3]      | 0.7899    | 
b[4]      | 0.7813    | 
b[5]      | 0.7852    | 
b[6]      | 0.7878    | 
b[7]      | 0.7974    | 
b[8]      | 0.7704    | 
b[9]      | 0.7611    | 
b[10]     | 0.7917    | 
b[11]     | 0.7827    | 
b[12]     | 0.7811    | 
b[13]     | 0.7903    | 
b[14]     | 0.7874    | 
b[15]     | 0.7907    | 
b[16]     | 0.7624    | 
b[17]     | 0.7859    | 
b[18]     | 0.7854    | 
b[19]     | 0.775     | 
b[20]     | 0.7854    | 
b[21]     | 0.7791    | 
b[22]     | 0.7817    | 
b[23]   

In [84]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 100 -leaf 500 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.7948    | 
b[2]      | 0.8018    | 
b[3]      | 0.8124    | 
b[4]      | 0.8108    | 
b[5]      | 0.7811    | 
b[6]      | 0.8109    | 
b[7]      | 0.7922    | 
b[8]      | 0.8049    | 
b[9]      | 0.806     | 
b[10]     | 0.8001    | 
b[11]     | 0.8167    | 
b[12]     | 0.7966    | 
b[13]     | 0.7998    | 
b[14]     | 0.8144    | 
b[15]     | 0.8163    | 
b[16]     | 0.791     | 
b[17]     | 0.8023    | 
b[18]     | 0.8061    | 
b[19]     | 0.8144    | 
b[20]     | 0.802     | 
b[21]     | 0.7947    | 
b[22]     | 0.8006    | 
b[23]   

In [85]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 500 -leaf 1000 -metric2t NDCG@20 -sparse -silent


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
b[1]      | 0.8256    | 
b[2]      | 0.8263    | 
b[3]      | 0.8432    | 
b[4]      | 0.8395    | 
b[5]      | 0.8402    | 
b[6]      | 0.8429    | 
b[7]      | 0.8462    | 
b[8]      | 0.828     | 
b[9]      | 0.8343    | 
b[10]     | 0.8387    | 
b[11]     | 0.8381    | 
b[12]     | 0.8401    | 
b[13]     | 0.8519    | 
b[14]     | 0.8365    | 
b[15]     | 0.8259    | 
b[16]     | 0.84      | 
b[17]     | 0.8479    | 
b[18]     | 0.8432    | 
b[19]     | 0.8281    | 
b[20]     | 0.8417    | 
b[21]     | 0.8385    | 
b[22]     | 0.84      | 
b[23]   

b[312]    | 0.8361    | 
b[313]    | 0.855     | 
b[314]    | 0.8312    | 
b[315]    | 0.8444    | 
b[316]    | 0.8386    | 
b[317]    | 0.8448    | 
b[318]    | 0.8352    | 
b[319]    | 0.8297    | 
b[320]    | 0.8298    | 
b[321]    | 0.8401    | 
b[322]    | 0.8426    | 
b[323]    | 0.8336    | 
b[324]    | 0.8598    | 
b[325]    | 0.8502    | 
b[326]    | 0.8324    | 
b[327]    | 0.8321    | 
b[328]    | 0.8318    | 
b[329]    | 0.8488    | 
b[330]    | 0.8368    | 
b[331]    | 0.8473    | 
b[332]    | 0.8357    | 
b[333]    | 0.855     | 
b[334]    | 0.8377    | 
b[335]    | 0.8425    | 
b[336]    | 0.8475    | 
b[337]    | 0.8564    | 
b[338]    | 0.8437    | 
b[339]    | 0.8455    | 
b[340]    | 0.8385    | 
b[341]    | 0.8468    | 
b[342]    | 0.8393    | 
b[343]    | 0.8524    | 
b[344]    | 0.8325    | 
b[345]    | 0.8449    | 
b[346]    | 0.84      | 
b[347]    | 0.8433    | 
b[348]    | 0.8444    | 
b[349]    | 0.8369    | 
b[350]    | 0.8493    | 
b[351]    | 0.8168    | 


# Статистика

In [87]:
!time java -jar ../../lib/RankLib-2.1-patched.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 0 -metric2t NDCG@20 -sparse -silent -save mart.txt


[+] General Parameters:
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	MART
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No
Model file: mart.txt

[+] MART's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7415

Model saved to: mart.txt
342.10user 1.33system 4:25.26elapsed 129%CPU (0avgtext+0avgdata 844764maxresident)k
229384inputs+4216outputs (4major+349620minor)pagefaults 0swaps


In [124]:
!time java -jar ../../lib/RankLib-2.12.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -metric2t NDCG@20 -sparse -silent -save random-forest.txt


Discard orig. features
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No
Model file: random-forest.txt

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7456

Model saved to: random-forest.txt
634.32user 5.16system 2:50.09elapsed 375%CPU (0avgtext+0avgdata 634436maxresident)k
240inputs+13600outputs (3major+237059minor)pagefaults 0swaps


In [112]:
!java -cp ../../lib/RankLib-2.10.jar:../../lib/commons-math3-3.6.1.jar ciir.umass.edu.features.FeatureManager -feature_stats mart.txt


Model File: /home/alexander/Desktop/information-retrieval-course/machine-learning/1/mart.txt
Algorithm : MART

Feature frequencies : 
	Feature[1] :       8
	Feature[2] :      22
	Feature[3] :      70
	Feature[4] :      47
	Feature[5] :      61
	Feature[6] :       2
	Feature[7] :      68
	Feature[8] :      48
	Feature[9] :      40
	Feature[10] :      76
	Feature[11] :      67
	Feature[12] :       3
	Feature[13] :      34
	Feature[14] :      34
	Feature[15] :     131
	Feature[16] :      27
	Feature[18] :      44
	Feature[20] :       3
	Feature[21] :      17
	Feature[22] :      14
	Feature[23] :      12
	Feature[24] :      32
	Feature[25] :      26
	Feature[26] :     170
	Feature[27] :      37
	Feature[28] :     148
	Feature[30] :      46
	Feature[31] :      73
	Feature[32] :      37
	Feature[33] :      67
	Feature[34] :       9
	Feature[35] :      48
	Feature[36] :      29
	Feature[37] :      17
	Feature[38] :     100
	Feature[39] :       3
	Feature[40] :      73
	Feature[43] :      45


In [125]:
!java -cp ../../lib/RankLib-2.10.jar:../../lib/commons-math3-3.6.1.jar \
           ciir.umass.edu.features.FeatureManager \
           -feature_stats random-forest.txt


Model File: /home/alexander/Desktop/information-retrieval-course/machine-learning/1/random-forest.txt
Algorithm : Random Forests

Feature frequencies : 
	Feature[1] :      41
	Feature[2] :     142
	Feature[3] :     334
	Feature[4] :     911
	Feature[5] :     208
	Feature[6] :       5
	Feature[7] :     349
	Feature[8] :     108
	Feature[9] :     137
	Feature[10] :     233
	Feature[11] :     520
	Feature[12] :      15
	Feature[13] :      86
	Feature[14] :     230
	Feature[15] :     518
	Feature[16] :     173
	Feature[17] :       1
	Feature[18] :     273
	Feature[19] :      18
	Feature[20] :      10
	Feature[21] :      66
	Feature[22] :      21
	Feature[23] :      17
	Feature[24] :     162
	Feature[25] :     118
	Feature[26] :     427
	Feature[27] :     155
	Feature[28] :     285
	Feature[30] :     279
	Feature[31] :     243
	Feature[32] :      71
	Feature[33] :     363
	Feature[34] :      21
	Feature[35] :      88
	Feature[36] :     102
	Feature[37] :      61
	Feature[38] :     171
	Fea

In [114]:
!time java -jar ../../lib/RankLib-2.12.jar -train ../../data/task3_train.txt -tts 0.2 -ranker 8 -bag 500 -leaf 10 -metric2t NDCG@20 -sparse -silent -save random-forest-smaller.txt


Discard orig. features
Training data:	../../data/task3_train.txt
Train-Test split: 0.2
Feature vector representation: Sparse.
Ranking method:	Random Forests
Feature description file:	Unspecified. All features will be used.
Train metric:	NDCG@20
Test metric:	NDCG@20
Feature normalization: No
Model file: random-forest-smaller.txt

[+] Random Forests's Parameters:

Reading feature file [../../data/task3_train.txt]... [Done.]            
(7033 ranked lists, 75057 entries read)
NDCG@20 on test data: 0.7337

Model saved to: random-forest-smaller.txt
913.18user 2.80system 4:20.85elapsed 351%CPU (0avgtext+0avgdata 637292maxresident)k
288inputs+2440outputs (1major+228316minor)pagefaults 0swaps


In [116]:
!java -cp ../../lib/RankLib-2.12.jar:../../lib/commons-math3-3.6.1.jar \
           ciir.umass.edu.features.FeatureManager \
           -feature_stats random-forest-smaller.txt


Model File: /home/alexander/Desktop/information-retrieval-course/machine-learning/1/random-forest-smaller.txt
Algorithm : Random Forests

Feature frequencies : 
	Feature[1] :       1
	Feature[2] :      23
	Feature[3] :      98
	Feature[4] :     258
	Feature[5] :      68
	Feature[7] :     111
	Feature[8] :       1
	Feature[9] :      20
	Feature[10] :      43
	Feature[11] :     311
	Feature[12] :       2
	Feature[13] :       5
	Feature[14] :      37
	Feature[15] :       5
	Feature[16] :      16
	Feature[18] :      43
	Feature[24] :     173
	Feature[25] :      60
	Feature[26] :      32
	Feature[27] :       4
	Feature[28] :       7
	Feature[30] :      78
	Feature[31] :       3
	Feature[32] :       1
	Feature[33] :     112
	Feature[34] :       4
	Feature[35] :       2
	Feature[37] :      12
	Feature[40] :      43
	Feature[43] :     110
	Feature[46] :      29
	Feature[47] :       2
	Feature[50] :      18
	Feature[51] :       8
	Feature[53] :       7
	Feature[54] :       4
	Feature[56] :    